Add GitHub Artifacts API Route

HassanAbouelela opened the pull request:
## Description
This PR adds an API route, which uses a [github app](https://docs.github.com/en/developers/apps/getting-started-with-apps/about-apps) (not to be confused with OAuth apps) to return a download URL for the static preview artifact. This should allow us to get around the rate limiting issues that we've been experiencing in #728.

**A bit more info on my approach:**
Q. Why a GitHub app?
A. It provides the most granularity in terms of permissions. We can get very specific about which resources we need, what type of access (read, write), and on which repos it's installed. It requires more work in the codebase than something like PATs, but it's about a million times safer.

Q. Why on the site, instead of a cloudflare worker?
A. The approach used here would've been the same had we used CF workers, but keeping it in site and keeping it in python makes it easier to debug and maintain as we have a lot more Python developers who can help keep the script running.

## Testing the PR
This PR is expected to fail actions, since it relies on a (till this is merged) non-existent API route. I've deployed the site, and tested everything out ([PR](https://github.com/HassanAbouelela/pydis-site/pull/4), [site](https://app.netlify.com/sites/clever-klepon-54aabb/deploys/)). I've taken down the site for now, but I can throw it back up if someone needs it for testing. The netlify build script does not need to be tested *on* netlify, if you don't want to.

## Roll-out plan
As mentioned, I expect the action to fail on this PR. Once the branch is ready, I'll update the command and environment variables in netlify, then we'll merge and hopefully see some green ticks. We also need to roll out the changes to bot-core, but that can be done in a separate PR.

HassanAbouelela committed:
Add GitHub Artifact API View

Adds an API route to fetch GitHub build artifacts through a GitHub app.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Update Netlify Build Script

Update the netlify build script to use the artifacts API from the site.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Make Awaiting Workflow Run A User Responsibility

Moves the responsibility of re-requesting a workflow run from the API to
the user. This makes the requests much shorter-lived, and allows the
client to control how they want to handle sleeping and retrying.

This also has the benefit of removing the only real piece of async code,
so now the view is completely sync once again.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Clean Up Artifact Tests

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

ChrisLovering commented on pydis_site/apps/api/github_utils.py:
```diff
@@ -0,0 +1,184 @@
+        raise e
+
+
+def check_run_status(run: dict) -> str:
```
Would it be a good idea to dataclass/namedtuple for this run arg? Just so that we can know what to expect in the data, and maybe error at creation time if it's missing a field we expect?

HassanAbouelela replied:
I think that sounds reasonable, though the class would be pretty bare bones because the documentation is too. I'd probably just include the fields we do use because it's also a very large model.

HassanAbouelela replied:
37001bc

ChrisLovering reviewed (approved):
From your testing it clearly works, the tests seem to cover quite a bit too.

Just have a question, that could be a nice to have/not needed, but shouldn't block this PR.

HassanAbouelela committed:
Merge branch 'main' into github-artifacts

HassanAbouelela committed:
Use Dataclass For Workflow Run

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Use UTC Time For GitHub API

When reading the created_at time from the GitHub API, it'll be a naive
date string with UTC time, so we use that instead of the system's time.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

jchristgit commented on pydis_site/apps/api/github_utils.py:
```diff
@@ -0,0 +1,209 @@
+    def from_raw(cls, data: dict[str, typing.Any]):
+        """Create an instance using the raw data from the API, discarding unused fields."""
+        return cls(**{
+            key.name: data[key.name] for key in dataclasses.fields(cls)
```
this is such a cool trick

jchristgit commented on pydis_site/apps/api/github_utils.py:
```diff
@@ -0,0 +1,209 @@
+        )
+
+    except BaseException as e:
+        # Close the client if we encountered an unexpected exception
+        client.close()
```
Shouldn't this be in a `finally`? Otherwise we leak, unless httpx closes it in `__del__`.

jchristgit replied:
This is also how it's done further below

HassanAbouelela replied:
The client only needs to be closed if we encounter an error. If we don’t, the client is returned to the caller to be used for all further interactions. The close below is what closes it in case of success.

jchristgit commented on pydis_site/apps/api/tests/test_github_utils.py:
```diff
@@ -0,0 +1,285 @@
+            """
+            Intercept the encode method.
+
+            The result is encoded with an algorithm which does not require a PEM key, as it may
```
is that why the addition of `*.pem` in `.gitignore`?

HassanAbouelela replied:
Yeah private key encoding is mandated for GitHub apps. In prod, we can deploy it as an environment variable, but this is more convenient for local testing.

jchristgit reviewed (approved):
This looks good. But shouldn't the endpoint have some authorization / whitelisting / ratelimiting?

HassanAbouelela committed:
Merge branch 'main' into github-artifacts

HassanAbouelela committed:
Rename GitHub App Environment Variables

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>
